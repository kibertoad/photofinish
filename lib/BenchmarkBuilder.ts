import {
  BenchmarkResults,
  executeBenchmarkAsync,
  executeBenchmarkSync,
} from './benchmarkExecutioner'

export type FunctionType = () => any
export type AsyncFunctionType = () => Promise<any>

export type Benchmark = {
  name: string
  warmupCycles: number // How many times function will be run before benchmarking starts in order to allow Node.js to perform optimizations
  benchmarkCycles: number // How many execution cycles are executed in order to collect samples
  benchmarkCycleSamples: number // How many samples per cycle are generated. Note that samples below 15 and above 85 percentile are discarded as insignificant
  functionUnderTest: FunctionType | undefined
  asyncFunctionUnderTest: AsyncFunctionType | undefined
  execute: () => BenchmarkResults
  executeAsync: () => Promise<BenchmarkResults>
}

export class BenchmarkBuilder {
  private _name = 'Benchmark'
  private _warmupCycles = 1000
  private _benchmarkCycles = 20000
  private _benchmarkCycleSamples = 100
  private _functionUnderTest: (() => any) | undefined
  private _asyncFunctionUnderTest: (() => Promise<any>) | undefined

  name(name: string): BenchmarkBuilder {
    this._name = name
    return this
  }

  warmupCycles(warmupCycles: number): BenchmarkBuilder {
    this._warmupCycles = warmupCycles
    return this
  }

  benchmarkCycles(benchmarkCycles: number): BenchmarkBuilder {
    this._benchmarkCycles = benchmarkCycles
    return this
  }

  benchmarkCycleSamples(benchmarkCycleSamples: number): BenchmarkBuilder {
    this._benchmarkCycleSamples = benchmarkCycleSamples
    return this
  }

  functionUnderTest(
    functionUnderTest: FunctionType,
    isPromiseResultAllowed = false
  ): BenchmarkBuilder {
    if (this._asyncFunctionUnderTest) {
      throw new Error(
        'asyncFunctionUnderTest is already set, please only set one of asyncFunctionUnderTest and functionUnderTest'
      )
    }

    const sanityCheckResult = functionUnderTest()
    if (!isPromiseResultAllowed && isPromise(sanityCheckResult)) {
      throw new Error(
        `Please use .asyncFunctionUnderTest() method for benchmarking async functions. If you are sure about what you are doing and don't want to await returned promise within benchmark, please set "isPromiseResultAllowed" parameter to true`
      )
    }

    this._functionUnderTest = functionUnderTest
    return this
  }

  asyncFunctionUnderTest(asyncFunctionUnderTest: AsyncFunctionType): BenchmarkBuilder {
    if (this._functionUnderTest) {
      throw new Error(
        'functionUnderTest is already set, please only set one of asyncFunctionUnderTest and functionUnderTest'
      )
    }

    this._asyncFunctionUnderTest = asyncFunctionUnderTest
    return this
  }

  build(): Benchmark {
    if (!this._asyncFunctionUnderTest && !this._functionUnderTest) {
      throw new Error('Either asyncFunctionUnderTest or functionUnderTest needs to bet set.')
    }

    const benchmark: Benchmark = {
      name: this._name,
      warmupCycles: this._warmupCycles,
      benchmarkCycles: this._benchmarkCycles,
      benchmarkCycleSamples: this._benchmarkCycleSamples,
      asyncFunctionUnderTest: this._asyncFunctionUnderTest,
      functionUnderTest: this._functionUnderTest,
      execute: () => {
        throw new Error('Function under test is asynchronous, use executeAsync() instead.')
      },
      executeAsync: () => {
        throw new Error('Function under test is synchronous, use execute() instead.')
      },
    }

    if (benchmark.functionUnderTest) {
      benchmark.execute = (): BenchmarkResults => {
        return executeBenchmarkSync(benchmark)
      }
    }

    if (benchmark.asyncFunctionUnderTest) {
      benchmark.executeAsync = (): Promise<BenchmarkResults> => {
        return executeBenchmarkAsync(benchmark)
      }
    }

    return benchmark
  }
}

function isPromise(value: any) {
  return Promise.resolve(value) === value
}
