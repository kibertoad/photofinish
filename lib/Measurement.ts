export class Measurement {
  private readonly timeInNanoSeconds: number
  constructor(timeInNanoSeconds: number) {
    this.timeInNanoSeconds = timeInNanoSeconds
  }

  getTimeInNanoSeconds(): number {
    return this.timeInNanoSeconds
  }

  getTimeInMilliSeconds(): number {
    return this.timeInNanoSeconds / 1000000
  }

  getTextInMsecs(precision = 3): string {
    return `${this.getTimeInMilliSeconds().toFixed(precision)} msecs`
  }

  getOpsPerSec(): number {
    return 1000000000 / this.getTimeInNanoSeconds()
  }

  getTextOpsPerSec(): string {
    return this.getOpsPerSec().toFixed(3)
  }
}
