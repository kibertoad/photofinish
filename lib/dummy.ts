export function dummy(): void {
  console.log('hi')
}
