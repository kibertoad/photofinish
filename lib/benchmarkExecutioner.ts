import { Benchmark, FunctionType } from './BenchmarkBuilder'

export type BenchmarkResults = {
  avgTime: number
}

async function warmupAsync(warmupCycles: number, functionUnderTest: FunctionType) {
  for (let i = 0; i < warmupCycles; i++) {
    await functionUnderTest()
  }
}

function warmup(warmupCycles: number, functionUnderTest: FunctionType) {
  for (let i = 0; i < warmupCycles; i++) {
    functionUnderTest()
  }
}

export function executeBenchmarkSync(benchmark: Benchmark): BenchmarkResults {
  if (!benchmark.functionUnderTest) {
    throw new Error('Function under test is not set')
  }
  warmup(benchmark.warmupCycles, benchmark.functionUnderTest)

  return {
    avgTime: 0,
  }
}

export async function executeBenchmarkAsync(benchmark: Benchmark): Promise<BenchmarkResults> {
  if (!benchmark.asyncFunctionUnderTest) {
    throw new Error('Async function under test is not set')
  }
  await warmupAsync(benchmark.warmupCycles, benchmark.asyncFunctionUnderTest)

  return {
    avgTime: 0,
  }
}
