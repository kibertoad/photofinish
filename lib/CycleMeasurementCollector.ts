import { Measurement } from './Measurement'

export type CycleMeasurements = {
  samples: number[]

  mean: Measurement
}

export class CycleMeasurementCollector {}
