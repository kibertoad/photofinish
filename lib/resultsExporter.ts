import { BenchmarkResults } from './internal/benchmarkExecutioner'
import path from 'path'
import fs from 'fs'

const nodeVersion = process.versions.node
const runtimeVersion = `${nodeVersion}, V8 ${process.versions.v8}`

export type TableData = {
  runtimeVersion: string
  benchmarkName: string
  benchmarkEntryName: string
  benchmarkEntryVersion?: string
  meanTimeNs: number
  meanTimeMs: number
}

export type ExportOptions = {
  exportPath?: string
}

export function exportResults(
  benchmarkResults: BenchmarkResults,
  options: ExportOptions = {}
): void {
  const exportPath = options?.exportPath ?? '/results'
  const exportData: TableData = {
    runtimeVersion,
    benchmarkName: benchmarkResults.benchmarkName,
    benchmarkEntryName: benchmarkResults.benchmarkEntryName,
    benchmarkEntryVersion: benchmarkResults.benchmarkEntryVersion,
    meanTimeNs: benchmarkResults.meanTime.getTimeInNanoSeconds(),
    meanTimeMs: benchmarkResults.meanTime.getTimeInMilliSeconds(),
  }

  const filename = `${benchmarkResults.benchmarkName}-${
    benchmarkResults.benchmarkEntryName
  }-Node_${nodeVersion.slice(0, 2)}.json`
  const resolvedPathDir = path.resolve(
    exportPath,
    normalizePath(benchmarkResults.benchmarkName),
    normalizePath(benchmarkResults.benchmarkEntryName)
  )
  const resolvedPathFile = path.resolve(resolvedPathDir, filename)

  fs.mkdirSync(resolvedPathDir, { recursive: true })
  fs.writeFileSync(resolvedPathFile, JSON.stringify(exportData, null, 2))
}

function normalizePath(path: string) {
  return path.replace(/ /g, '_').toLowerCase()
}
