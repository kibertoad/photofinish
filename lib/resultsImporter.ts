import { TableData } from './resultsExporter'

const { loadFiles } = require('load-goblin')

export async function loadResults(baseResultsDir: string): Promise<TableData[]> {
  const loadResults = await loadFiles({
    filter: '*.json',
    rootDirectory: baseResultsDir,
    recursively: true,
  })

  return loadResults.map((entry: Record<string, any>) => {
    return {
      runtimeVersion: entry.content.runtimeVersion,
      benchmarkName: entry.content.benchmarkName,
      benchmarkEntryName: entry.content.benchmarkEntryName,
      benchmarkEntryVersion: entry.content.benchmarkEntryVersion,
      meanTimeNs: entry.content.meanTimeNs,
      meanTimeMs: entry.content.meanTimeMs,
      warmupCycles: entry.content.warmupCycles,
      benchmarkCycles: entry.content.benchmarkCycles,
      benchmarkCycleSamples: entry.content.benchmarkCycleSamples,
      label: entry.content.label,
    }
  })
}
