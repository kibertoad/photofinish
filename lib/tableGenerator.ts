import { TableData } from './resultsExporter'
import { Measurement } from './Measurement'

const table = require('markdown-table')

export type TableOptions = {
  precision?: number
}

export function generateTable(results: readonly TableData[], options: TableOptions = {}): string {
  const tableIndex: string[] = ['Node', 'Option', 'Msecs/op', 'Ops/sec', 'V8']

  const tableRows: string[][] = [tableIndex]
  results.map((entry) => {
    const measurement = new Measurement(entry.meanTimeNs)
    const [nodeVersion, v8Version] = entry.runtimeVersion.split(', ')

    const row = [
      nodeVersion,
      entry.benchmarkEntryVersion
        ? `${entry.benchmarkEntryName} ${entry.benchmarkEntryVersion}`
        : entry.benchmarkEntryName,
      measurement.getTextInMsecs(options.precision ?? 3),
      measurement.getTextOpsPerSec(),
      v8Version,
    ]

    tableRows.push(row)
  })

  return table(tableRows)
}
