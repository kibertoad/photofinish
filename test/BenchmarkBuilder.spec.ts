import { BenchmarkBuilder } from '../lib/BenchmarkBuilder'

describe('BenchmarkBuilder', () => {
  it('creates correct benchmark instance', () => {
    const benchmarkBuilder = new BenchmarkBuilder()

    const benchmark = benchmarkBuilder
      .benchmarkName('Dummy name')
      .benchmarkEntryName('Only entry')
      .warmupCycles(500)
      .benchmarkCycles(44)
      .benchmarkCycleSamples(55)
      .functionUnderTest(() => {})

    expect(benchmark).toMatchSnapshot()
  })

  it('does not allow to set asyncFunctionUnderTest if functionUnderTest is set', () => {
    const benchmarkBuilder = new BenchmarkBuilder()

    expect(() => {
      benchmarkBuilder
        .benchmarkName('Dummy name')
        .benchmarkEntryName('Only entry')
        .warmupCycles(500)
        .benchmarkCycles(44)
        .benchmarkCycleSamples(55)
        .functionUnderTest(() => {})
        .asyncFunctionUnderTest(() => Promise.resolve())
    }).toThrow(/functionUnderTest is already set/)
  })

  it('does not allow to set functionUnderTest if asyncFunctionUnderTest is set', () => {
    const benchmarkBuilder = new BenchmarkBuilder()

    expect(() => {
      benchmarkBuilder
        .benchmarkName('Dummy name')
        .benchmarkEntryName('Only entry')
        .warmupCycles(500)
        .benchmarkCycles(44)
        .benchmarkCycleSamples(55)
        .asyncFunctionUnderTest(() => Promise.resolve())
        .functionUnderTest(() => {})
    }).toThrow(/asyncFunctionUnderTest is already set/)
  })
})
