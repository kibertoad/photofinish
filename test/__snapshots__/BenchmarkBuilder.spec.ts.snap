// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BenchmarkBuilder creates correct benchmark instance 1`] = `
BenchmarkBuilder {
  "_benchmarkCycleSamples": 55,
  "_benchmarkCycles": 44,
  "_benchmarkEntryName": "Benchmark entry",
  "_benchmarkName": "Dummy name",
  "_functionUnderTest": [Function],
  "_warmupCycles": 500,
}
`;
